python/tvm/micro/contrib/stm32/emitter.py

```diff
@@ -0,0 +1,1369 @@
+                f"""
+        // {DBAR}
+        //   ai_{name}_create
+        // {DBAR}
+        AI_API_ENTRY
+        ai_status ai_{name}_create(
+          const ai_ptr weights,
+          const ai_ptr activations
+        )
+        {{
+          ai_status status = AI_STATUS_OK;
+          status = {name}_configure_weights (weights);
+          if (status != AI_STATUS_OK) {{
+            return status;
+          }}
+          status = {name}_configure_activations (activations);
+          if (status != AI_STATUS_OK) {{
+            return status;
+          }}
+          return AI_STATUS_OK;
+        }}
+        """
+            )
+        )
+
+        out_c.write(
+            textwrap.dedent(
+                f"""
+        // {DBAR}
+        //   ai_{name}_destroy
+        // {DBAR}
+        AI_API_ENTRY
+        ai_status ai_{name}_destroy ()
+        {{
+          return AI_STATUS_OK;
+        }}
+        """
+            )
+        )
+
+    def emit_code(self, dest_dir, model_name):
+        """ Emits the C code implementing the model. """
+
+        # Build the directory structure
+        if os.path.exists(dest_dir):
+            raise ValueError(f"emit_code.Error: {dest_dir} exists.")
+
+        # Make a new one
+        os.makedirs(dest_dir)
+
+        # Fix the model name
+        model_name = re.sub("[^0-9a-zA-Z_]+", "_", model_name)
+        model_name = model_name.lower()
+
+        # Write the C code: we can parse the string
+        if isinstance(self.lib_, list):
+            # List of strings from Model Library Format C files
+            for idx, src in enumerate(self.lib_):
+                code = _preprocess_code(src)
+                filename = os.path.join(dest_dir, f"{model_name}_lib{idx}.c")
+                with open(filename, "w") as fout:
+                    fout.write(code)
+        else:
+            # a TVM RuntimeGraphFactory
+            src = self.lib_.get_source(fmt="c")
+            code = _preprocess_code(src)
+            filename = os.path.join(dest_dir, f"{model_name}_lib.c")
+            with open(filename, "w") as fout:
+                fout.write(code)
+
+        # Save params as binary data
+        saved_params = tvm.runtime.save_param_dict(self._params)
+        params_name = os.path.join(dest_dir, model_name + ".params")
+        with open(params_name, "wb") as f:
+            f.write(saved_params)
+
+        # Write the .json
+        graph_name = os.path.join(dest_dir, model_name + ".json")
+        json_string = json.dumps(self.graph_, indent=4)
+        with open(graph_name, "w") as f:
+            print(json_string, file=f)
+
+        # emit X_data[c,h]
+        data_h_name = os.path.join(dest_dir, model_name + "_data.h")
+        data_c_name = os.path.join(dest_dir, model_name + "_data.c")
+        model_h_name = os.path.join(dest_dir, model_name + ".h")
+        model_c_name = os.path.join(dest_dir, model_name + ".c")
+
+        data_h = open(data_h_name, "w")
```

Comment:
could you do this like so:

```
import contextlib
with contextlib.ExitStack() as exit_stack:
  data_h = exit_stack.enter_context(open(data_h_name, "w"))
  data_c = exit_stack.enter_context(open(data_c_name, "w"))
  out_h = exit_stack.enter_context(open(out_h_name, "w"))
  out_c = exit_stack.enter_context(open(out_h_name, "w"))

  self._emit_params_data(...)
  self.emit_open(...)
  # .. all other self._emit

# no need to close() the ExitStack does this for you
```

this way, if an exception occurs in any `self._emit_*`, Python will auto-close the files for you.